### DISCUSSION TITLE
OCPP 2.0.1 support

### DISCUSSION THREAD

**Discussion by munterkalmsteiner** [lbbrhzn/ocpp#983] [Q&A]
Are there any plans already to support OCPP 2.0.1? I'm planning to buy a [DEFA Power](https://www.defa.com/product/defa-power/) which implements v2.0.1.

**Comment by lbbrhzn** [comment]
No plans yet... perhaps we should. OCPP 2.0 is not backwards compatible with OCPP 1.6

**Reply by tobsys** [reply]
+1, that would be awesome.

**Comment by JoaoPedroBelo** [comment]
any news on this?

**Reply by magnushacker** [reply]
I have a DEFA power and can confirm that it does not work. Also, you need to contact their support to get the charger configured, there's no way to do it by yourself. 

The Python OCPP library claims to support 2.0.1 so why doesn't this integration?

**Comment by piomark** [comment]
any update/plans about ocpp 2.0.1 support?

I see more and more newly released chargers drop support for previous protocols and implement only 2.0.1

**Reply by bobcov** [reply]
Does OCPP exist because these devices don't speak MQTT?  Otherwise, could MQTT handle what this is doing with chargers, etc?

**Reply by piomark** [reply]
Well, those are completely different things. 
MQTT is communication protocol that defines pub/sub communication mechanism. It does not define what kind of information is transferred. 

OCPP is an application protocol (using REST/HTTP as communication protocol), and as such it defines how a charger talks with a charging station management system to control it (it defines all messages structure and data that is transferred in between).

**Comment by bobcov** [comment]
But there are charging stations which speak MQTT and seem to do just fine, no?

**Comment by magnushacker** [comment]
What kind of effort would be involved in adding support for ocpp 2.0.1? I understand some parts need to be reimplemented, but I cannot judge how much work would be needed. Perhaps it could be worthwhile reaching out the community for assistance in case the maintainers do not have time?

**Reply by piomark** [reply]
I created a fork of this project to see what can be done. I don't have much time right now therefore the progress is really slow. 
The only thing that I managed to do is to replace all references/data structures from 1.6 to 2.0.1. 
The charger is connecting to the server (so there is a small success) but no logic is implemented. 

Regarding the effort, it's a complete re-work of OCPP layer. 2.0.1 is completely different than 1.6 when it comes to messages and communication.

When (and if) I have time I will continue the work on the project. In my free time I'm digging through 2.0.1 spec.

**Reply by quokkapatrik** [reply]
What are your plan for letting version 1.6 co-exist with version 2.0.1?

**Reply by piomark** [reply]
For the moment it's proof of concept ;) I was not intending to preserve 1.6, I am working on a fork so nothing will change here. 
If I succeed then ofc it can be done, for the moment I have no clue when I will find sufficient time to do it. 
Anyway my goal is not to reproduce the functionality as in this repo. I am only interested in get/set charging current/power and get some basic statistics (at least for now).

**Comment by magnushacker** [comment]
Are there even chargers that talk both 1.6 and 2.0.1? Maybe there won't be
much of an issue if users had to choose between two separate HACS repos
depending on which charger they have.

**Comment by purcell-lab** [comment]
Looks like OCPP v2.0.1 support was added in v0.6.0
![image-6](https://github.com/user-attachments/assets/a42b382c-a1a6-4e95-bc29-54413c7721c4)
